<img style="float: center; width: 100%" src="https://raw.githubusercontent.com/andrejkk/TalksImgs/master/FrontSlideUpperBan.png">
<p style="margin-bottom:2cm;"></p>

<center>
    <H1> 9. Markovske verige </H1>
   


<br><br>
    <H3> Andrej Košir, Lucami, FE </H4>
    
    <H4> Kontakt: prof. dr. Andrej Košir, andrej.kosir@lucami.fe.uni-lj.si, skype=akosir_sid </H4>
</center>


<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 1 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> Cilji </div>
</div>


## ■ Cilji

- Cilj: 
    - Modeliranje z markovskimi verigami
    - Osnove analize stanj markovske verige


- Slučajni proces na končno stanjih, ki je najbolje poznan in uporaben. Omejimo se na:
    - Diskreten čas
    - Diskreten prostor stanj (končno ali števno neskončno stanj)


- Primer: uporabniki in njihovo obnašanje
- Primeri uporabe
    - Analiza končnega avtomata: analizaobnašanja TK sistema, ...
    - Ekonomski modeli


- Kaj nas zanima:
    - Pogostnost obiskov
    - Stanje po dolgem času
    - Pojavi kot je absorbcija


<figure>
<img style="float: center; width: 300px; margin: 0px 20px 20px 0px;" src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/MarkovChain_Emotions.png">
  <figcaption></figcaption>
</figure>

<figure>
<img style="float: right; width: 200px; margin: -320px 20px 20px 0px;" src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/MarkovChain_Marker.png">
  <figcaption></figcaption>
</figure>


<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 2 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> Cilji </div>
</div>


## ■ Poglavja 


9.1. Definicija, osnovne lastnosti

■ Definicija markovske verige $\large{*}$

■ O prehodni matriki, porazdelitev stanj $\large{*}$


9.2. Klasifikacija stanj

■ Povezanost stanj in perioda $\large{*}$

■ Povrnljivo in minljivo stanje $\large{*}$

■ Ergodijska mar. veriga in časi vračanja v stanje

■ Stacionarna (limitna) porazdelitev $\large{*}$

■ Končna markovska veriga $\large{*}$


9.3. Ocenjevanje prehodne matrike

■ Ocene prehodnih verjetnosti

■ Interval zaupanja in grafikov kvantilov

■ Določane velikosti vzorca

■ Numerični vidik

■ Primer: razredi uporabnikov TK storitev


<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 3 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.1. Definicija, osnovne lastnosti </div>
</div>


## 9.1. Definicija, osnovne lastnosti

■ Definicija markovske verige

■ O prehodni matriki, porazdelitev stanj


<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 4 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.1. Definicija, osnovne lastnosti </div>
</div>

## ■ Definicija markovske verige


- Prostor stanj $S$, $m=|S|$
    - Končen, neskončen (števen)
- Čas $T$
    - Diskreten, zvezen

- Slučajni proces: 
$$ X_n : T \to S $$

- Diskretna Markovska veriga je slučajni proces na diskretnem času in diskretnim prostorom stanj, za katerega velja pravilo prehodne verjetnosti enega koraka
$$ P[X_{n+1}|_{X_n, X_{n-1}, \ldots, X_0}] = P[X_{n+1}|_{X_n}] = P[X_1|_{X_0}] $$


- Prehodna verjetnost enega koraka
$$ p_{ij} = P[X_{n+1}=j|_{X_n=i}] $$


- Končen prostor stanj – matrika prehodnih verjetnosti = prehodna matrija
$$ P = [p_{ij}], \; i,j\in S. $$

Vrstične vsote ima enake 1.


<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 5 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

In [6]:
import numpy as np
from sympy import Matrix
from sympy.functions import re

# Case 1 
# Transition matrix
P = np.array([[0.7, 0.2, 0.1, 0.0], 
              [0.2, 0.5, 0.2, 0.1], 
              [0.05, 0.15, 0.6, 0.2], 
              [0.0, 0.2, 0.0, 0.8]])

P

array([[0.7 , 0.2 , 0.1 , 0.  ],
       [0.2 , 0.5 , 0.2 , 0.1 ],
       [0.05, 0.15, 0.6 , 0.2 ],
       [0.  , 0.2 , 0.  , 0.8 ]])

<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.1. Definicija, osnovne lastnosti </div>
</div>


## ■ O prehodni matriki, porazdelitev stanj


- Prehodna matrika za $n$ korakov
    - $p_{ij}^{(n)} = P[X_{k+n}=j|_{X_k=i}]$, $P^{(n)}=[p_{ij}^{(n)}]$
    - Formula o popolni verjetnosti
    $$ P^{(n)} = P^n $$
    
    
   
- Porazdelitev stanj: 
  $$ \pi_i^n = P[X_n=i], \pi^n = [\pi_1^n, \ldots, \pi_i^n] $$
    - To je delež časa sistema v danem stanju na dolgi rok? Ne vedno.
    - Začetna porazdelitev: $\pi^0$. Če je stanje $X_0=i_0$ znano, je         
    $$ \pi^0=\delta_{i_0 i} $$ 
    - Formula o popolni verjetnosti:
    $$ \pi^n = P^{(n)} \pi^0 $$
    - O limitni porazdelitvi kasneje

<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 6 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.2. Klasifikacija stanj </div>
</div>


## 9.2. Klasifikacija stanj


■ Povezanost stanj in perioda

■ Povrnljivo in minljivo stanje

■ Ergodijska mar. veriga in časi vračanja v stanje

■ Stacionarna (limitna) porazdelitev

■ Končna markovska veriga


<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 7 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.2. Klasifikacija stanj </div>
</div>


## ■ Povezanost stanj in perioda


- Dosegljivo stanje
    - Stanje $j$ je dosegljivo iz stanja $i$, če $\exists n: p_{ij}^{(n)}>0$
    - Stanji sta povezani, če sta vzajemno dosegljivi, označimo $i\leftrightarrow j$
    - Povezanost je ekvivalenčna relacija


- Markovska veriga je nerazcepna, če ima en sam razred stanj;
    - Prehodna matrika je bločno diagonalna


- Perioda stanja $d(i)$ je največji skupni delitelj vseh $n$, za katere velja $P_{ii}^{(n)} > 0$ 
    - Če je $i\leftrightarrow j$, potem $d(i) = d(j)$;
    - Stanje je neperiodično, če je $d(i)=1$;
    
    
- Stanje je ničelno, če je   
$$ \lim_{n\to\infty} P_{ii}^{(n)} = 0. $$


<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 8 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

In [7]:
# ---------------------------------------------------------------------------------------------------
# Classification of states
print ('Transition matrix P=\n', P)
print ('\nTransition matrix P^2=\n', np.linalg.matrix_power(P, 2))

# All entries of P^2 are non-zero
# => All atates are connected

Transition matrix P=
 [[0.7  0.2  0.1  0.  ]
 [0.2  0.5  0.2  0.1 ]
 [0.05 0.15 0.6  0.2 ]
 [0.   0.2  0.   0.8 ]]

Transition matrix P^2=
 [[0.535 0.255 0.17  0.04 ]
 [0.25  0.34  0.24  0.17 ]
 [0.095 0.215 0.395 0.295]
 [0.04  0.26  0.04  0.66 ]]


In [22]:
# ---------------------------------------------------------------------------------------------------
# Decompose transition matrix
# Use Jordan decomposition
#mP = Matrix(P)
#mB, mJ = mP.jordan_form()
#B, J = np.array(mB), np.array(mJ)
B = np.array([[-0.5,-0.776313,0.659526,-0.413722],
              [-0.5,-0.212847,-0.11514,0.830469],
              [-0.5,0.283275,-0.73692,-0.123153],
              [-0.5,0.521335,0.0933625,-0.352121]])
J = np.array([[1.,0,0,0],
              [0,0.718345,0,0],
              [0,0,0.553349,0],
              [0,0,0,0.328305]])

# Test decomposition
err_mat = P - np.dot(B, np.dot(J, np.linalg.inv(B)))
print ('\nDecomposition error norm = ', np.linalg.norm(err_mat))


Decomposition error norm =  7.249920892047682e-07


<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.2. Klasifikacija stanj </div>
</div>


## ■ Povrnljivo in minljivo stanje


- Verjetnost prve vrnitve po $n$ korakih
$$ f_{ij}^n = P[X_n=j|_{X_{n-1}\ne j, \ldots, X_1\ne j, X_0=i}] $$


- Slučajna spremenljivka števila korakov prehoda iz $i$ v $j$ označimo s $T_{ij}$ in ima porazdelitev 
$$ T_{ij} \sim \left(\matrix{0 & 1 & 2 & \ldots \cr f_{ij}^0 & f_{ij}^1 & f_{ij}^2 & \ldots}\right) $$
    - Posebej označimo $T_i = T_{ii}$
    - Označimo: $\sum_{n=0}^\infty f_{ii}^n = f_{ii}^*$. Vrednost 
    $$ 1- f_{ii}^* $$
    je verjetnost, da se nikoli več ne vrnemo v stanje $i$ potem ko smo stanje $i$ zapustili.

- Stanje $i$ je **povrnljivo**, če je $f_{ii}^* = 1$. Stanje je **minljivo**, če ni povrnljivo.
    - Obe lastnosti se ohranjata pri povezanosti, torej definirata ekvivalenčne razrede.

- Velja: stanje $i$ povrnljivo je natanko tedaj, ko je
$$ \sum_{n=1}^\infty P_{ii}^{(n)} = \infty $$


- Velja : če $i\leftrightarrow j$, sta obe stanji ali minljivi ali povarnljivi


<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 9 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

In [23]:
# Recurent, transient? 

# Eigenvalues
eigVals, eigVecs = np.linalg.eig(P.T)

# Get it sorted
idx = eigVals.argsort()[::-1]   
eigVa = eigVals[idx]
eigVc = eigVecs[:,idx]

las_bf =  (1./(1-eigVa[1:4]))-1
print ('las_bf', las_bf)
           
sumJn = np.diag(np.insert(las_bf, 0, np.inf))
print ('\nSum J^n\n', sumJn)

sumPn = np.dot(B, np.dot(sumJn, np.linalg.inv(B)))
print ('\nSum P^n\n', sumPn)

# Since all diagonal elements of sum P^n are infinity
#  => all states are recurrent


las_bf [2.55044912 1.23888596 0.48877142]

Sum J^n
 [[       inf 0.         0.         0.        ]
 [0.         2.55044912 0.         0.        ]
 [0.         0.         1.23888596 0.        ]
 [0.         0.         0.         0.48877142]]

Sum P^n
 [[inf inf inf inf]
 [inf inf inf inf]
 [inf inf inf inf]
 [inf inf inf inf]]


In [24]:
def inf_sum(a_lst):
    return [(la/(1.0-la) if la < 1 else np.infty) for la in a_lst]

# ---------------------------------------------------------------------------------------------------
# Are states recurrent?
# Compute sum of P^n
sumJ = np.diag(inf_sum(np.diag(J)))
sumP = np.dot(B, np.dot(sumJ, np.linalg.inv(B)))

print ('\nSum of P^n = \n', sumP)
# => All states are reccurent 


Sum of P^n = 
 [[inf inf inf inf]
 [inf inf inf inf]
 [inf inf inf inf]
 [inf inf inf inf]]


<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.2. Klasifikacija stanj </div>
</div>


## ■ Ergodijska mar. veriga in časi vračanja v stanje


- Markovska veriga je Ergodijska, če je
    - Nerazcepna
    - Povrnljiva (vsa stanja so povrnljiva)
    - Neperiodična (vsa stanja so nepriodična)
    - Neničelna (vsa stanja so neničelna)
    
    
- Pričakovan čas vračanja: 
$$ E(T_i) = \sum_{n=0}^\infty n f_{ii}^{(n)} $$


- Za vsa stanja Ergodijske markovske verige velja
$$ \lim_{n\to\infty} P_{ii}^{(n)} = \frac{1}{E(T_i)} $$
$$ \lim_{n\to\infty} P_{ij}^{(n)} = \lim_{n\to\infty} P_{ii}^{(n)} $$ 



- Povrnljivo **ničelno** stanje $i$
$$ \lim_{n\to\infty} P_{ii}^{(n)} = 0 \; \Leftrightarrow \;  E(T_i) = \infty  $$


- **Povrnljivo pozitivno** = **ergodijsko stanje** $i$: 
$$ \lim_{n\to\infty} P_{ii}^{(n)} > 0 \; \Leftrightarrow \; E(T_i) < \infty  $$


<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 10 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

In [25]:
def inf_lim(a_lst):
    return [(0 if la < 1 else 1) for la in a_lst]

# ---------------------------------------------------------------------------------------------------
# Zero or positive states?
# Compute lim of P^n
limJ = np.diag(inf_lim(np.diag(J)))
limP = np.dot(B, np.dot(limJ, np.linalg.inv(B)))

print ('\nlim of P^n = \n', limP)

# Since all lim P^n are positive
# => all states are positive

# Expected times of visits
ETi = 1.0/np.diag(limP)

print ('\nExpected number of visits py states:\n', ETi)



lim of P^n = 
 [[0.21301781 0.27218919 0.18934915 0.32544385]
 [0.21301781 0.27218919 0.18934915 0.32544385]
 [0.21301781 0.27218919 0.18934915 0.32544385]
 [0.21301781 0.27218919 0.18934915 0.32544385]]

Expected number of visits py states:
 [4.69444311 3.67391519 5.28124903 3.07272668]


<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.2. Klasifikacija stanj </div>
</div>


## ■ Stacionarna (limitna) porazdelitev


- Spomnimo se 
$$ \pi_i^n = P[X_n=i], \qquad \pi^n = [\pi_1^n, \ldots, \pi_i^n] $$


- Stacionarna (limitna) porazdelitev: 
  $$ \pi = \lim_{n\to\infty} \pi^n $$
    - Enolično določena: vsa stanja so povrnljiva neničelna


- Za ergodijsko markovsko verigo obstaja limitna porazdelitev
- Velja: limitna porazdelitev Ergodijske Markovske verige je levi lastni vektor prehodne matrike:
  $$ \pi^\top = \pi^\top P. $$



- Opazimo: V ergodijsk verigi je limitna porazdelitev neodvisna od začetne porazdelitve $\pi^0$.


<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 11 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

In [26]:
# ---------------------------------------------------------------------------------------------------
# Limit distribuion
eigVals, eigVecs = np.linalg.eig(P.T)
pi = eigVecs[:, 0]/sum(eigVecs[:, 0])

print ('\neigVals=\n', eigVals)
#print '\neigVecs=\n', eigVecs
print ('\nLimit distribution =\n', pi)


eigVals=
 [1.         0.32830522 0.71834549 0.55334929]

Limit distribution =
 [0.21301775 0.27218935 0.18934911 0.32544379]


<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.2. Klasifikacija stanj </div>
</div>



## ■ Verjetnost absorbcije


- Absorbcija: veriga dane podmnožice stanj $S_0 \subset S$ ne zapusti več 
- Verjetnost absorpcije: verjetnost, da se to zgodi
- Če je veriga na začetku v minljivem stanje, se vedno zgodi:
    1. Slučajni proces z verjetnostjo 1 zapusti množico minljivih stanj
    2. Slučajni proces z verjetnostjo 1 ostane v množici povrnljivih stanj, v katera je vstopil;


- Absorbcijsko stanje $i$: edino stanje povrnljivega razreda. 
    Velja: $i$ absorbirajoče če in samo če $P_{ii}=1$.
    
    
- Naj bo $i\in S$ minljivo stanje, $S_e$ razred povrnljivih stanj: <br>
**Verjetnost absorbcije:** $\pi_i(S_e)$ je verjetnost, da je veriga z $X_0=i$ absorbirana v $S_e$ 


- Velja – formula za izračun - za vsak $i$
$$ \lim_{n\to\infty} P_{ij}^{(n)} = \pi_i(S_e)\pi_j. $$


<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 12 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.2. Klasifikacija stanj </div>
</div>


## ■ Končna markovska veriga

- Ima končno množico stanj $|S|=m < \infty$.
- Vsaka končna Markovska veriga stanja vsebuje tudi povrnljiva stanja.
- Vsa povrnljiva stanja so pozitivna, torej **ergodijska** in imajo končen pričakovan čas vrnitve
$$ \lim_{n\to\infty} P_{ii}^{(n)} > 0. $$


- Klasifikacija stanj: vsa stanja so $S=S_m\cup S_e$, kjer je 
    - Minljiva stanja: $S_m$
    - Ergodijska (povrnljiva neničelna): $S_e$
    
    
- Bločna razdelitev prehodne matrike: uredimo indekse stanj
    - Ergodisjka stanja: $1, \ldots, m-K$
    - Minljiva stanja: $m-K+1, \ldots, m$
    $$ P = \left[\matrix{S & 0 \cr R & Q}\right] $$
    
    
- Fundamentalna matrika: 
    $$ N = (I - Q)^{-1}.  $$


<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 13 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

In [2]:
import numpy as np
# Case 2
c2P = np.array(
    [[0.7, 0.3, 0.0, 0.0], 
     [0.2, 0.8, 0.0, 0.0], 
     [0.05,0.15, 0.6, 0.2], 
     [0.0, 0.2, 0.0, 0.8]])
print ('Transition matrix P=\n', c2P)


Transition matrix P=
[[0.7  0.3  0.   0.  ]
 [0.2  0.8  0.   0.  ]
 [0.05 0.15 0.6  0.2 ]
 [0.   0.2  0.   0.8 ]]


In [3]:
# ---------------------------------------------------------------------------------------------------
# Classification of states
print ('Transition matrix P=\n', c2P)
print ('\nTransition matrix P^2=\n', np.linalg.matrix_power(c2P, 2))
print ('\nTransition matrix P^3=\n', np.linalg.matrix_power(c2P, 3))

# Matrix multiplications by blocks
# => 2 x 2 block of zeros stayes the same
# => There are two connected classes of states


Transition matrix P=
[[0.7  0.3  0.   0.  ]
 [0.2  0.8  0.   0.  ]
 [0.05 0.15 0.6  0.2 ]
 [0.   0.2  0.   0.8 ]]

Transition matrix P^2=
[[0.55  0.45  0.    0.   ]
 [0.3   0.7   0.    0.   ]
 [0.095 0.265 0.36  0.28 ]
 [0.04  0.32  0.    0.64 ]]

Transition matrix P^3=
[[0.475  0.525  0.     0.    ]
 [0.35   0.65   0.     0.    ]
 [0.1375 0.3505 0.216  0.296 ]
 [0.092  0.396  0.     0.512 ]]


In [4]:
# ---------------------------------------------------------------------------------------------------
# Decompose transition matrix
# Use Jordan decomposition
#mP = Matrix(c2P)
#mB, mJ = mP.jordan_form()
#B, J = np.array(mB), np.array(mJ)
c2B = np.array([[1.0, 0, 0, 9.0/4],
              [1.0, 0, 0, -3.0/2],
              [1.0, 1.0, 1.0, -7.0/8],
              [1.0, 1.0, 0, 1.0]])
c2J = np.array([[1.0,0,0,0],
              [0,4.0/5,0,0],
              [0,0,3.0/5,0],
              [0,0,0,1.0/2]])

# Test decomposition
err_mat = c2P - np.dot(c2B, np.dot(c2J, np.linalg.inv(c2B)))
print ('\nDecomposition error norm = ', np.linalg.norm(err_mat))


Decomposition error norm =  1.9675159943996247e-16


In [5]:
def prd(a,b):
    if np.isinf(a) and b == 0:
        return 0
    if np.isinf(b) and a == 0:
        return 0
    return a*b

def myDot(A,B):
    nA1, nA2, nB2 = A.shape[0], A.shape[1], B.shape[1]
    AB = np.zeros((nA1,nB2))
    for i in range(nA1):
        for j in range(nB2):
            curr = 0
            for k in range(nA2):
                curr += prd(A[i,k], B[k,j])
            AB[i, j] = curr
    return AB

# Classification of states
# Recurent, transient? 

# Eigenvalues
eigVals, eigVecs = np.linalg.eig(c2P.T)

# Get it sorted
idx = eigVals.argsort()[::-1]   
eigVa = eigVals[idx]
eigVc = eigVecs[:,idx]

las_bf =  (1./(1-eigVa[1:4]))-1
sumJn = np.diag(np.insert(las_bf, 0, np.inf))

sumPn = myDot(c2B, myDot(sumJn, np.linalg.inv(c2B)))
print ('\nSum P^n\n', sumPn)

# Since diagonal elements of sum P^n i=1 and i=2 are infinity
#  => states i=1 and i=2 are recurrent
# Since diagonal elements of sum P^n i=3 and i=4 are finite
#  => states i=3 and i=4 are transient



Sum P^n
[[inf inf 0.  0. ]
 [inf inf 0.  0. ]
 [inf inf 1.5 2.5]
 [inf inf 0.  4. ]]


In [6]:
def inf_lim(a_lst):
    return [(0 if la < 1 else 1) for la in a_lst]

# ---------------------------------------------------------------------------------------------------
# Zero or positive states?
# Compute lim of P^n
limJ = np.diag(inf_lim(np.diag(c2J)))
limP = np.dot(c2B, np.dot(limJ, np.linalg.inv(c2B)))

print ('\nlim of P^n = \n', limP)

# Since all lim P^n are positive
# => all states are positive

# Expected times of visits
ETi = 1.0/np.diag(limP)

print ('\nExpected number of visits py states:\n', ETi)


lim of P^n = 
[[0.4 0.6 0.  0. ]
 [0.4 0.6 0.  0. ]
 [0.4 0.6 0.  0. ]
 [0.4 0.6 0.  0. ]]

Expected number of visits py states:
[2.5        1.66666667        inf        inf]


/home/nbuser/anaconda2_501/lib/python2.7/site-packages/ipykernel/__main__.py:16: RuntimeWarning: divide by zero encountered in divide


In [7]:
# ---------------------------------------------------------------------------------------------------
# Limit distribuion
eigVals, eigVecs = np.linalg.eig(c2P.T)


# Get it sorted
idx = eigVals.argsort()[::-1]   
eigenValues = eigVals[idx]
eigenVectors = eigVecs[:,idx]

# Normalize eigenvector
pi = eigenVectors[:, 0]/sum(eigenVectors[:, 0])

print ('\neigVals=\n', eigenValues)
#print '\neigVecs=\n', eigenVectors
print ('\nLimit distribution =', pi)


eigVals=
[1.  0.8 0.6 0.5]

Limit distribution = [ 0.4  0.6 -0.  -0. ]


<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.2. Klasifikacija stanj </div>
</div>



## ■ Število obiskov minljivega stanja

- Slučajna spremenljivka $n_i$: število obiskov stanja $i$ v poljubno dolgem času
- Matematično upanje števila obiskov $j$ če je začetno stanje $i$: 
 $$ E_i[n_j] < \infty $$


- Velja: če je stanje $i$ minljivo, je 
 $$ E_i[n_j] = \sum_{k=1}^\infty P_{ij}^{(k)} $$


- Velja: fundamentalna matrika vsebuje pričakovana števila vračanj:
$$ \left[E_i[n_j]\right] = N $$ 



<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 14 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

In [8]:
# Expected number of visits of recurent states
m = 4 # |S|
K = 2 # number of recurent states

Q = c2P[K:m, K:m]
print ('Part of transition matrix Q = \n', Q)

N = np.linalg.inv(np.eye(m-K)-Q)
print ('\nFundamental matrix N =\n', N)

Part of transition matrix Q = 
[[0.6 0.2]
 [0.  0.8]]

Fundamental matrix N =
[[2.5 2.5]
 [0.  5. ]]


<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.3 Ocenjevanje prehodne matrike </div>
</div>


## 9.3 Ocenjevanje prehodne matrike

■ Ocene prehodnih verjetnosti

■ Interval zaupanja in grafikov kvantilov

■ Določane velikosti vzorca

■ Numerični vidik

■ Primer: razredi uporabnikov TK storitev


<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 15 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.3 Ocenjevanje prehodne matrike </div>
</div>



## ■ Ocene prehodnih verjetnosti

- Preodna verjetnost
$$ p_{ij} = \frac{n_k}{n} $$
je statistična ocena;


- Ocena parametra (npr. povprečje ali delež): 
    - ocenimo in določimo
        - povprečje - pričakovano vrednost: $\overline{x}$
        - standardni odklon: $\sigma$
        - za stopnjo tveganja $\alpha$ določimo $z_\alpha$ (za normalno porazdelitev je $z_{0.05}=1.96$)
    - Interval zaupanja: 
    $$ CI = [\overline{x} - z_\alpha\frac{\sigma}{\sqrt{n}}, \overline{x} + z_\alpha\frac{\sigma}{\sqrt{n}}] $$
    - Velikost vzorca: določimo ga glede na predpisano velikost intervala zaupanja
    $$ |CI| = 2 z_\alpha\frac{\sigma}{\sqrt{n}} $$


<img style="float: right; width: 250px; margin: 0px 20px 20px 0px;" src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/EstimationConfInterval.png">



<p style="margin-bottom:7cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 16 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.3 Ocenjevanje prehodne matrike </div>
</div>



## ■ Interval zaupanja in grafikov kvantilov


- Vizualizacija intervala zaupanja

<img style="float: center; width: 350px; margin: 0px 20px 20px 0px;" src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/EstimationConfIntervalVsSampleSize.png">



<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 17 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.3 Ocenjevanje prehodne matrike </div>
</div>



## ■ Določane velikosti vzorca

- Iz zahteve po širini intervala zaupanja
- Graf velikosti vzorca za oceno deleža $\alpha=0.05$ ($z_\alpha=1.96$)
$$ CI = [p_0 - z_\alpha\sqrt{p_0(1-p_0)/n}, p_0 + z_\alpha\sqrt{p_0(1-p_0)/n}] $$


- Za primer $p=0.65$ in $|CI|=\Delta p$ pri $\alpha=0.05$ velja neenačba
$$ n\geq \frac{4 z_\alpha^2}{\Delta p^2} p_0 (1-p_0) $$

- Od tod dobimo:
    - Za $\Delta p=0.01:$ $n\geq 34959$,
    - Za $\Delta p=0.02:$ $n\geq 175$,
    - Za $\Delta p=0.04:$ $n\geq 88$.


<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 18 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.3 Ocenjevanje prehodne matrike </div>
</div>



## ■ Numerični vidik - matrični račun


- Potrebujemo izračun spektra in potenc 
$$ P^n = P^{(n)} $$


- Jordanova dekompozicija
$$ P = B J B^{-1},  $$
kjer je $J$ Jordanova kanonična forma.


- Za vsak razred stanj posebej velja
$$ J = \lambda I + R $$


- Velja
$$ P^n = B J^n B^{-1} $$


<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 19 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.3 Ocenjevanje prehodne matrike </div>
</div>



## ■ Primer: razredi uporabnikov TK storitev


- Množica uporabnikov TK storitve $U$
- Razdelimo jih po zadovoljstvu: $U=U_1 \cup U_2 \cup U_3 \cup U_4$
    - Zelo zadovoljni: $U_1$
    - Zadovoljni: $U_2$
    - Nezadovoljni: $U_3$
    - Zapuščajo ponudnika: $U_4$


- Začetna porazdelitev: ocena trenutnega stanja uporabnikov
- Rezultati:
    - Povrnljivost - minljivost stanj in razredi teh stanj
    - Limitna porazdelitev



<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 20 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">

<div style="display:flex;font-weight:bold;font-size:0.9em;">
<div style="flex:1;width:50%;"> 9. Markovske verige  </div>
<div style="flex:1;width:50%;text-align:right;"> 9.3 Ocenjevanje prehodne matrike </div>
</div>



## ■ Zaključek

- Preprosta in uporabna analiza končnih stanj
- Preverjanje pogojev pri oceni začetne porazdelitve in prehodne matrike

<p style="margin-bottom:2cm;"></p>
<div style="width:100%;text-align:right;font-weight:bold;font-size:1.2em;"> 21 </div>
<img src="https://raw.githubusercontent.com/andrejkk/ORvTK_SlidesImgs/master/footer_full.jpg">